test/test_data/map.rb

```diff
@@ -0,0 +1,61 @@
+# typed: true
+# frozen_string_literal: true
+
+class TestGenerics
+  extend T::Sig
+
+  sig { params(should_succeed: T::Boolean).returns(Typed::Result[Integer, String]) }
+  def do_something(should_succeed)
+    if should_succeed
+      Typed::Success.new(123)
+    else
+      Typed::Failure.new("")
+    end
+  end
+
+  sig { params(should_succeed: T::Boolean).returns(Typed::Result[String, Integer]) }
+  def do_something_else(should_succeed)
+    if should_succeed
+      Typed::Success.new("success")
+    else
+      Typed::Failure.new(5)
+    end
+  end
+
+  sig { params(should_succeed: T::Boolean).returns(Typed::Result[Time, Float]) }
+  def do_one_more_thing(should_succeed)
+    if should_succeed
+      Typed::Success.new(Time.now)
+    else
+      Typed::Failure.new(3.14)
+    end
+  end
+
+  sig { returns(Typed::Result[T::Boolean, T.any(String, Integer, Float)]) }
+  def test_map
+    do_something(true)
+      .map { |payload| do_something_else(T.assert_type!(payload, Integer) == 123) }
+      .map { |payload| do_one_more_thing(T.assert_type!(payload, String) == "success") }
+      .map { |payload| Typed::Success.new(T.assert_type!(payload, Time) == Time.now) }
```

Comment:
Question: when a failure occur here, it falls through to the end, correct? So we start ignoring the blocks and carry the same failure all the way down?

Reply:
Correct, as soon as a `Typed::Failure` is returned by a block, the whole chain breaks and the failure is returned